### PR TITLE
fixes #445 Rule Verification doesn't show any result after executing

### LINKED ISSUE #445: Rule Verification doesn't show any result after executing
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the CDP4-IME
- [x] I have searched [open](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
After executing a RuleVerificationList I would expect some kind of result and logging of the verification. 
I also expect the status of a User Rule Verification to change after execution. At the moment it doesn't.

Investigate and fix. Create new issue if it takes more time.

### PR DESCRIPTION
**Pull request by SimonWoodRhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Problem:
Violation rows were not being displayed against the validation rules when running the rule.
This was happening for the following reasons:

1) Violations are not persistent and don't update the revision number therefore updates were being ignored by the view model reactive query.

2) Violations were being overwritten when the rule was modified and written to the data store.

3) The view was not displaying the description text (just empty rows)

Solution:
Add the violations after the validation rule has been persisted. Listen for changes to the validation rule in the view model that do not have revision changes.

The solution looks a little convoluted because of the need to sequence the updates and listen to modifications.

### CODE REVIEW

File: CDP4Composition/Services/RuleVerification/RuleVerificationService.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@ namespace CDP4Composition.Services
     using System;
     using System.Collections.Generic;
     using System.ComponentModel.Composition;
+    using System.Reactive.Linq;
```
Comment (samatrhea) [inline CDP4Composition/Services/RuleVerification/RuleVerificationService.cs:12]: when you update a source-file, feel free to update the copyright header with the one referencing GPL

File: CDP4Composition/CommonView/AutoGenRows/RuleViolationRowViewModel.cs [hunk 2]
```diff
@@ -77,4 +84,4 @@ private void UpdateProperties()
             this.Description = this.Thing.Description;
         }
     }
-}
+}
```
Comment (lxatrhea) [inline CDP4Composition/CommonView/AutoGenRows/RuleViolationRowViewModel.cs:87]: Every code file should end with a line break.

**Review by samatrhea — CHANGES_REQUESTED** [review]
please refactor the event on the Name property to make use of ObservableAsPropertyHelper

**Review by samatrhea — CHANGES_REQUESTED** [review]
please refactor the Name property using the ReactiveUI with ObservableAsPropertyHelper pattern, read more about it here https://www.reactiveui.net/docs/handbook/observable-as-property-helper/

File: CDP4Composition/Services/RuleVerification/RuleVerificationService.cs [hunk 3]
```diff
@@ -246,19 +251,35 @@ private void Execute(ISession session, UserRuleVerification userRuleVerification
...
+
+                                        CDPMessageBus.Current.SendObjectChangeEvent(verification, EventKind.Updated);
+
+                                        foreach (var ruleViolation in violations)
+                                        {
+                                            CDPMessageBus.Current.SendObjectChangeEvent(ruleViolation, EventKind.Added);
+                                        }
+                                    });
             }
+
```
Comment (lxatrhea) [inline CDP4Composition/Services/RuleVerification/RuleVerificationService.cs:281]: I think there should be an else {} implementation where userRuleVerification.Status is set to `RuleVerificationStatusKind.Passed`.

File: EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs [hunk 4]
```diff
@@ -35,6 +38,14 @@ protected RuleVerificationRowViewModel(RuleVerification ruleVerification, ISessi
             : base(ruleVerification, session, containerViewModel)
         {
             this.UpdateProperties();
+
+            //A subscription is made here in addition to the one made in the base class in order to be notified of non-persisntent changes which will not result in an updated revision number
```
Comment (lxatrhea) [inline EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs:42]: small typo persisntent => persistent

File: EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs [hunk 5]
```diff
@@ -35,6 +38,14 @@ protected RuleVerificationRowViewModel(RuleVerification ruleVerification, ISessi
...
+            var thingSubscription = CDPMessageBus.Current.Listen<ObjectChangedEvent>(this.Thing)
```
Comment (lxatrhea) [inline EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs:43]: There is a virtual method `InitializeSubscriptions` in the RowViewModelBase that you could override where these kind of subscriptions can be created.

File: EngineeringModel/Views/RuleVerificationListBrowser/RuleVerificationListBrowser.xaml [hunk 6]
```diff
@@ -80,7 +80,7 @@
                         </Image>
                     </dx:PixelSnapper>
                     <ContentPresenter x:Name="defaultRowPresenter"
-                                      Content="{Binding Description, Mode=OneWay, UpdateSourceTrigger=PropertyChanged}"
+                                      Content="{Binding}"
```
Comment (lxatrhea) [inline EngineeringModel/Views/RuleVerificationListBrowser/RuleVerificationListBrowser.xaml:83]: Is this change as expected?

**Review by lxatrhea — CHANGES_REQUESTED** [review]
Some small changes requested. Next to that, I think some headers should also be updated.

File: EngineeringModel/Views/RuleVerificationListBrowser/RuleVerificationListBrowser.xaml (hunk 6 above)
Comment (SimonWoodRhea) [reply EngineeringModel/Views/RuleVerificationListBrowser/RuleVerificationListBrowser.xaml:83]: The previous code appears to be redundant I therefore replaced it with the code in the above templates. The description is now exposed though a name property which is bound to from the tree grid control.

File: CDP4Composition.Tests/RuleVerification/RuleVerificationServiceTestFixture.cs [hunk 7]
```diff
@@ -10,14 +10,16 @@ namespace CDP4Composition.Tests.RuleVerification
     using System.Collections.Concurrent;
     using System.Collections.Generic;
     using System.Linq;
-    using System.Reactive.Concurrency;    
+    using System.Reactive.Concurrency;
```
[the remark this answers is not in the document]
Comment (SimonWoodRhea) [reply CDP4Composition.Tests/RuleVerification/RuleVerificationServiceTestFixture.cs:13]: Done.

File: CDP4Composition/Services/RuleVerification/RuleVerificationService.cs (hunk 3 above)
Comment (SimonWoodRhea) [reply CDP4Composition/Services/RuleVerification/RuleVerificationService.cs:304]: Done

File: EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs (hunk 5 above)
Comment (SimonWoodRhea) [reply EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs:48]: Done

**Review by samatrhea — APPROVED** [review]
looks good to me, feel free to land

File: EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs (hunk 4 above)
Comment (SimonWoodRhea) [reply EngineeringModel/ViewModels/RuleVerificationListBrowser/RuleVerificationRowViewModel.cs]: Done

**Review by lxatrhea — APPROVED** [review]

**Comment by alexatrhea** [comment]
Congrats on first merge @SimonWoodRhea!

**Pull request merged**